doc/dualtor/ICMP_Hardware_Offload_and_Protecion.md

```diff
@@ -0,0 +1,289 @@
+
+Currently switching time after link state change is detected, is around 25-50 milliseconds. Note this depends on route scale. Requirement for FRR protection switching using Nexthop Protection Group is to keep it consistent irrespective of the route scale.
+
+### High Level Component and Requirements
+#### LinkMgrd Requirements
+LinkMgrd is the central component that runs in MUX docker and is responsible for managing ICMP echo sessions and link state. Link prober sub-component in Linkmgrd is responsible for the ICMP echo sessions.
+* Requirements
+   * Add / Remove Hardware ICMP echo session in App DB based on mux cable and link prober config entries.
+   * Consume ICMP echo session state from State DB produced by Orchagent and update mux state.
+   * Create two hardware sessions per mux port / interface one for self and another one to monitor the peer replies.
+   * Generate and handle ICMP packets with TLVs in software.
+
+Following diagram shows the hardware ICMP echo sessions between the TORs and Server.
+<div align="center"> <img src=image/hw_icmp_echo_sessions.png width=600 /> </div>
+
+#### Orchagent Requirement
+Currently orchagent creates tunnel at initialization and add / removes routes to forward traffic to peer ToR via a IpinIP tunnel when linkmgrd switches state to standby / active.
+* Requirements
+   * Create / Remove hardware ICMP echo sessions by consuming entries in App DB.
+   * Consume ICMP echo session notification from SAI and update session state in State DB.
+   * Create Protection Next Hop group for FRR switchover based on config.
+   * Initiate traffic switchover by toggling members of Protection Next Hop group by consuming link state produced by LinkMgrd.
+
+#### SaiRedis Requirement
+A new SAI specification was added to support ICMP Echo offload sessions in SAI version 1.12.
+* Requirements
+  * Support all SAI attributes of sai icmp echo session from saiicmpecho.h.
+  * Support SAI switch notification for sai icmp echo sessions.
+
+## Detailed Design
+Following diagram describes high level interaction of components in SONiC.
+<div align="center"> <img src=image/sonic_component_interaction.png width=700 /> </div>
+
+### LinkMgrd
+With introduction of ICMP hardware offload a new config knob **[link_prober_type]** is added to differentiate between software and hardware based link probing. Hardware based Link Prober is added in LinkMgrd to support hardware based probing that produces the ICMP echo sessions entries in App DB ICMP_ECHO_SESSION_TABLE that gets consumed by orchagent. LinkMgrd will consume icmp echo session entries from State DB ICMP_ECHO_SESSION_TABLE to determine the hardware prober state and trigger events for Link state machines.
+
+#### Session cookie
+Session cookie is used by NPU to distinguish between software ICMP packets and ICMP packets that needs to be handled by the offload engine in NPU. LinkMgrd will use separate cookies for software and hardware based sessions.
+
+#### Session GUID handling

```

Comment:
Maybe also added a reference to the packet formatting too. Got offline questions about where this GUID is placed.  https://github.com/opencomputeproject/SAI/blob/0c11fa0d4f623440d0c5d575fec3ae0fb14947d7/doc/SAI-Proposal-ICMP-ECHO.md#220-icmp-echo-payload-packet